src/Modules/Trinity.DynamicCluster/Trinity.DynamicCluster/ChunkedStorage.cs

```
@@ -2,52 +2,183 @@
+                    Log.WriteLine(LogLevel.Error, ex.ToString());
+                    return TrinityErrorCode.E_FAILURE;
+                }
+            }
+            return TrinityErrorCode.E_SUCCESS;
         }
 
         public override TrinityErrorCode AddCell(long cellId, byte* buff, int size, ushort cellType)
         {
-            throw new NotImplementedException();
+            foreach (Storage.Storage s in PickStorages(a => Chunks[a].Covers(cellId)))
+            {
+                try
+                {
+                    s.AddCell(cellId, buff, size, cellType);
+                }
+                catch (Exception ex)
+                {
+                    Log.WriteLine(LogLevel.Error, "Errors occurred during AddCell.");
+                    Log.WriteLine(LogLevel.Error, ex.ToString());
+                    return TrinityErrorCode.E_FAILURE;
+                }
+            }
+            return TrinityErrorCode.E_SUCCESS;
         }
 
         public override TrinityErrorCode AddCell(long cellId, byte[] buff, int offset, int size, ushort cellType)
         {
-            throw new NotImplementedException();
+            foreach (Storage.Storage s in PickStorages(a => Chunks[a].Covers(cellId)))
+            {
+                try
+                {
+                    s.AddCell(cellId, buff, offset, size, cellType);
+                }
+                catch (Exception ex)
+                {
+                    Log.WriteLine(LogLevel.Error, "Errors occurred during AddCell.");
+                    Log.WriteLine(LogLevel.Error, ex.ToString());
+                    return TrinityErrorCode.E_FAILURE;
+                }
+            }
+            return TrinityErrorCode.E_SUCCESS;
         }
 
         public override bool Contains(long cellId)
         {
-            throw new NotImplementedException();
+            bool f = false;
+            foreach (Storage.Storage s in PickStorages(a => Chunks[a].Covers(cellId)))
```

Please consider rewrite these routines with LINQ queries, like:
```
return PickStorages(cellId).Where(TrinityErrorCode.E_SUCCESS == _AddCell_impl(...)); // don't forget that we already have "PickStorages" overloaded! Do it DRY.
// for contains:
return PickStorages(cellId).Where(s => s.Contains(cellId));
```

Also, don't swallow the exceptions. Let them propagate.
Think about it, why don't I swallow it inside storage.Contains?
Like, if there is a networking error, and an IOException comes out, what will you do?
Can we simply break, return E_FAILURE?
Or, what can we do to ensure delivery to all the chunks?